### ISSUE TITLE
Have `not_null<>` support function pointers and `std::function<>`

### ISSUE THREAD

**Issue by mbeutel** [gsl-lite/gsl-lite#233]
Cf. https://github.com/microsoft/GSL/issues/626.

**Comment by petamas** [comment]
I agree that it would be nice to have, and not only for function pointers (like microsoft/GSL#626), but std::function too (like microsoft/GSL#325).

Only question is if we want a whole new specialization for it, SFINAE `operator()`, or simply implement the operator to work when `T` is callable. (See my notes in #207)
